common/hdr-model.cpp

```diff
@@ -0,0 +1,345 @@
+}
+
+void hdr_model::apply_hdr_config()
+{
+    if( ! _device.is< rs2::serializable_device >() )
+        throw std::runtime_error( "Device not serializable" );
+    auto serializable = _device.as< rs2::serializable_device >();
+    serializable.load_json( _changed_config.to_json() );
+    _current_config = _changed_config;
+}
+
+void hdr_model::render_control_item( hdr_config::control_item & control, int /*idx*/ )
+{
+    ImGui::PushID( &control );
+
+    ImGui::Text( "Depth Gain:" );
+    ImGui::SameLine();
+    ImGui::SetNextItemWidth( 100 );
+    int g = control.DepthGain;
+    if( ImGui::InputInt( "##gain", &g ) )
+        control.DepthGain = std::max( 1, std::min( g, 16384 ) );
+
+    ImGui::Text( "Depth Exposure:" );
+    ImGui::SameLine();
+    ImGui::SetNextItemWidth( 100 );
+    int e = control.DepthManExp;
+    if( ImGui::InputInt( "##exp", &e ) )
+        control.DepthManExp = std::max( 1, std::min( e, 200000 ) );
+
+    ImGui::PopID();
+}
+
+void hdr_model::render_preset_item( hdr_config::preset_item & item, int idx )
+{
+    ImGui::PushID( idx );
+    std::string hdr = "Preset Item " + std::to_string( idx + 1 );
+    if( ImGui::CollapsingHeader( hdr.c_str() ) )
+    {
+        ImGui::Text( "Iterations:" );
+        ImGui::SameLine();
+        ImGui::SetNextItemWidth( 100 );
+        int it = item.iterations;
+        if( ImGui::InputInt( "##it", &it ) )
+            item.iterations = std::max( 1, it );
+
+        ImGui::Separator();
+        ImGui::Text( "controls:" );
+        if( ! item.controls.empty() )
+            render_control_item( item.controls[0], 0 );
+    }
+    ImGui::PopID();
+}
+
+void hdr_model::render_hdr_config_window( ux_window & window, std::string & error_message )
+{
+    if( ! _window_open || ! _hdr_supported )
+        return;
+
+    ImGui::SetNextWindowSize( { 680, 600 }, ImGuiCond_FirstUseEver );
+    if( ! ImGui::Begin( "HDR Configuration", &_window_open, ImGuiWindowFlags_NoCollapse ) )
+    {
+        ImGui::End();
+        return;
+    }
+
+    if( ! error_message.empty() )
+        ImGui::TextColored( { 1, 0.3f, 0.3f, 1 }, "%s", error_message.c_str() );
+    ImGui::Separator();
+
+    // SubPreset config
+    if( ImGui::CollapsingHeader( "SubPreset Configuration" ) )
+    {
+        ImGui::Text( "Preset ID:" );
+        ImGui::SameLine();
+        ImGui::SetNextItemWidth( 150 );
+        char buf[32];
+        strncpy( buf, _changed_config.SubPreset.id.c_str(), 31 );
+        buf[31] = '\0';
+        if( ImGui::InputText( "##id", buf, 32 ) )
+            _changed_config.SubPreset.id = buf;
+
+        ImGui::Text( "Global Iterations:" );
+        ImGui::SameLine();
+        int gi = _changed_config.SubPreset.iterations;
+        ImGui::SetNextItemWidth( 150 );
+        if( ImGui::InputInt( "##gi", &gi ) )
+            _changed_config.SubPreset.iterations = std::max( 0, gi );
+    }
+
+    // Preset Items
+    if( ImGui::CollapsingHeader( "Preset Items", ImGuiTreeNodeFlags_DefaultOpen ) )
```

Comment:
Should not be collapsible. All internal objects are collapsible and there is no differentiation. Confusing.